### ISSUE TITLE
Network IDs should be consecutive

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#4677]
### Describe the project you are working on

A arcade spaceship game

### Describe the problem or limitation you are having in your project

I need a way to index things with player id's, however the networking signals don't happen in the correct order to setup something like this.

for example if I want to index a node's children with a network id I can't, neither can I compare node names with it.

right now network id's are used to identify a peer however they can be so much more useful if they were consecutive...
for example spawning players...
```
func spawn_player(id)->Node:
   $PlayersSpawns.get_children()[id].add_child(player_scene.instantiate())

func destroy_player(id)->void:
    $PlayersSpawns.get_children()[id].get_children()[0].queue_free()
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Network id's should be consecutive so they count like this...
`1,2,3,4`

But they should also take in account disconnected players.
for example if id 2 disconnects the id that would be 4 would be 2
```
add 1
add 2
add 3
remove 2
add 2
add 4
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Simply put that id's would work the same way they would just have different values

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I have been trying to but it seems due to the order and unreliability of how peers connects its impossible

### Is there a reason why this should be core and not an add-on in the asset library?

Its already in core this is just a improvement of it.

**Comment by fire** [comment]
Why should they not be unique GUIDs?

**Comment by Calinou** [comment]
In a system where connections and disconnections can happen in an unpredictable manner, GUIDs are often preferable to incremental keys.

**Comment by dsnopek** [comment]
I'm not sure this would be an improvement, especially the part about reusing peer ids after a peer has disconnected. For example, if peer id 2 disconnects, and another peer connects immediately and gets assigned peer id 2, there's a chance (depending on timing and the underlying transport mechanism) that the new peer 2 could receive messages intended for the old peer 2. This could lead to weird intermittent bugs that are impossible to reproduce.

I don't what exactly the problem was that you encountered, but it *should* be possible to assign your own sequential player numbers to the players, without the peer ids needing to be sequential. This is something I've setup in all my multiplayer games.

**Comment by Shadowblitz16** [comment]
I mean just because they are reused doesn't mean they aren't unique ids. 
And the game devs should handle whether a user disconnects or not anyways.

I guess simply having a peer index implemented would fix the issue.
I don't necessarily need it to be the id I just need something to index things with that works in this way.

**Comment by YuriSizov** [comment]
If you need to index connections, you can already do that by creating some array or slotted dictionary structure from the existing GUIDs. This is perfectly solvable in scripts to fit your design.

**Comment by Shadowblitz16** [comment]
> If you need to index connections, you can already do that by creating some array or slotted dictionary structure from the existing GUIDs. This is perfectly solvable in scripts to fit your design.

Can you give a example? I don't think that works because they have to be ordered and reused.

**Comment by YuriSizov** [comment]
Then order and reuse them. Have a dictionary with consecutive numeric keys, and fill it in order with actual connection IDs. When a connection is no longer tracked, clear the indexed key in the dictionary, but don't remove it. The next time you get a new connection, consider empty keys first before adding a new connection to the end.
